### ISSUE TITLE
Cloaking and uncloaked addresses for specific clients or staff-only servers

### ISSUE THREAD

**Issue by Robby-** [inspircd/inspircd#1408]
I have a fairly unique setup in where I run several inspircd instances for specific purposes, but they are linked to each other, though. This provides me with some more flexibility and benefits that I wouldn't otherwise have when running everything off of one inspircd instance.

So, I run instances for our own webchat only, for IRC clients only, and then for staff only.

We run a few bots to monitor some things and to set bans (using CIDR blocks, etc.) on channels (in cases where setting an xline is not needed yet). These bots are connected to the staff instance/server and depend on being able to see normal, uncloaked, hostnames and IP addresses to do their thing, so we cannot enable cloaking.

Even though I used bots as an argument above, it is also easier on the staff to do their work and enabling cloaking would complicate things.

I do however wish to enable it at some point but for this I want to make the following feature request first:
Have the ability for some designated clients to see uncloaked addresses in JOIN/PART/QUIT/and-whatever-else-I'm-forgetting. Or, maybe instead of designated clients, even a whole designated server, such as the staff server, where everything appears uncloaked in JOIN/PART/etc for any client/staff connected to it, BUT still retain the possibility of clients/staff on such a designated server to enable cloaking for their own addresses, of course.

Also, not everyone is an OPER, in fact most are not and are just regular channel operators, so the latter proposed method seems most ideal for us as opposed to the first proposed method where it would probably have to depend on OPER and a privilege to determine if the client is allowed to see uncloaked addresses and I don't like to have to create an oper block for everyone.

Hell, maybe we can even have both methods? So that we and anyone else that finds this useful for their network, can choose which method to use.

**Comment by del6597** [comment]
When given an OPER block, aren't you able to view the uncloaked IP address with a WHOIS on the user?

**Comment by Robby-** [comment]
Yes, but that is besides the point, I need uncloaked addresses in JOIN/PART/QUIT/etc as I described in my opening post.

**Comment by Robby-** [comment]
As talked about on IRC a while ago, this approach (having a designated server to show all uncloaked addresses to connected clients) would open up a whole new can of worms. For example, uncloaked channel bans set by staff or bots that are connected to these designated servers would have to be automatically converted by the ircd to a cloaked form when sent out to the other servers. This is because channel mode changes aren't hidden from normal users and they would thus still be able to see the uncloaked bans being set by staff otherwise.

Note, however, that depending on *how* this would ever get implemented, for example with no auto-converting being done at all, `insp3` introduces a new `hidemode` module that could be used to work around this. It allows hiding specific channel mode changes from users that are below a certain rank. However, not all networks may wish to do this, and also, as it is currently implemented it is a network-wide thing, not per channel, so individual channels can't opt-in or opt-out of this. And this approach in itself in turn brings its own can of worms, as a channel operator (or otherwise sufficiently ranked user) connected to a normal server would still be able to see the uncloaked bans being set by staff, so an auto-conversion approach is still the preferred method.
